### DISCUSSION TITLE
Using conditional timelines to present randomized, not fixed timelines?

### DISCUSSION THREAD

**Discussion by LangDevLab** [jspsych/jsPsych#702] [Q&A] [answered]
Hi,

In our experiment, participants are assigned to one of three conditions (condition 4, 5 or 6) at the beginning of the study (their condition is based on their results from a separate experiment). Each condition shows 4 blocks of 8 trials that comprise of instructions, video clips and audio clips. The video and audio stimuli for each condition are different and the 4 blocks are randomized.

We read information on conditional timelines, but we're confused on how to implement the conditional timeline function for our experiment. We would like participants to enter their condition number that we assigned to them (either 4, 5 or 6) at the beginning of the experiment. They would then be presented with a timeline according to their condition. However, we would like blocks within each timeline to be randomized, rather than present a set block order. We already randomized the order of the blocks for each condition using the randomint function.

A snippet of how we randomized block order for condition 4 (it's messy but it works for the time being):

```js
// randomize block order // 
// order of one trial = 1) instruction trial 2) video trial 3) audio trial //

randomint = Math.floor(Math.random()*480);

if(randomint > 400){
  timeline.push(
    say_ABA, con4ABA_procedure, audio_procedureDR4,
    say_ABA, con4ABA_procedure, audio_procedureDR4,
    say_ABA, con4ABA_procedure, audio_procedureDR4,
    say_ABA, con4ABA_procedure, audio_procedureDR4,
    say_ABA, con4ABA_procedure, audio_procedureDR4,
    say_ABA, con4ABA_procedure, audio_procedureDR4,
    say_ABA, con4ABA_procedure, audio_procedureDR4,
    say_ABA, con4ABA_procedure, audio_procedureDR4,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    halfway,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    say_ADA, incon4ABA_procedure, audio_procedureDR4,
    say_ADA, incon4ABA_procedure, audio_procedureDR4,
    say_ADA, incon4ABA_procedure, audio_procedureDR4,
    say_ADA, incon4ABA_procedure, audio_procedureDR4,
    say_ADA, incon4ABA_procedure, audio_procedureDR4,
    say_ADA, incon4ABA_procedure, audio_procedureDR4,
    say_ADA, incon4ABA_procedure, audio_procedureDR4,
    say_ADA, incon4ABA_procedure, audio_procedureDR4
  );

} else if(randomint > 320){
  timeline.push(
    say_ADA, incon4ABA_procedure, audio_procedureDR4,
    say_ADA, incon4ABA_procedure, audio_procedureDR4,
    say_ADA, incon4ABA_procedure, audio_procedureDR4,
    say_ADA, incon4ABA_procedure, audio_procedureDR4,
    say_ADA, incon4ABA_procedure, audio_procedureDR4,
    say_ADA, incon4ABA_procedure, audio_procedureDR4,
    say_ADA, incon4ABA_procedure, audio_procedureDR4,
    say_ADA, incon4ABA_procedure, audio_procedureDR4,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    say_ABA, incon4ADA_procedure, audio_procedureDR4,
    halfway,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    say_ADA, con4ADA_procedure, audio_procedureDR4,
    say_ABA, con4ABA_procedure, audio_procedureDR4,
    say_ABA, con4ABA_procedure, audio_procedureDR4,
    say_ABA, con4ABA_procedure, audio_procedureDR4,
    say_ABA, con4ABA_procedure, audio_procedureDR4,
    say_ABA, con4ABA_procedure, audio_procedureDR4,
    say_ABA, con4ABA_procedure, audio_procedureDR4,
    say_ABA, con4ABA_procedure, audio_procedureDR4,
    say_ABA, con4ABA_procedure, audio_procedureDR4
  );
} 
```

Thank you, any help or insight is appreciated!

**Comment by jodeleeuw** [comment, marked as answer]
Since you are assigning users a condition from a previous experiment, would one option be to send a variable in the URL? For example, participants in condition 4 would go to:

http://mywonderfulexperiment.com/index.html?condition=4

While participants in condition 5 go to:

http://mywonderfulexperiment.com/index.html?condition=5

Then you can use [jsPsych.data.getURLVariable('condition')](https://www.jspsych.org/core_library/jspsych-data/#jspsychdatageturlvariable) to get the value of `condition`. Since this value is available when the page loads, you can use it when building your experiment. 

```js
var condition = jsPsych.data.getURLVariable('condition');

if(condition == 4){
 // ... add all the relevant bits to the timeline ...
}

if(condition == 5){
 // ... etc. ...
}
```

If this doesn't work then I can suggest an approach using the conditional function on timelines.

**Comment by LangDevLab** [comment]


**Reply by jodeleeuw** [reply]
When you tell them their condition for the next round of the experiment, could you send them a differentiated link at that point?

**Comment by LangDevLab** [comment]
I believe so. The way we have our experiment set up is participants email us when they have finished the first part of the experiment, we then send them a link to the second part of the experiment and tell them their condition.

edit: Ah sorry, I understand what you're saying now. That may work for us! Would it be possible to do so through JATOS as that is where our experiment is run.

**Reply by jodeleeuw** [reply]
I'm not certain about JATOS. Maybe @becky-gilbert would know? I believe she has some experience with JATOS.

**Reply by becky-gilbert** [reply]
JATOS also allows you to pass information into the experiment through the URL, which is what Josh was suggesting above with `jsPsych.data.getURLVariable('condition')`. When you send participants the link to the second part, instead of telling them their condition, you can just add it to the end of their study link. 

When you add the first URL parameter to a link, it has the form: ?parameter=value. However JATOS links already contain some URL parameters, which means that you'll be adding _another_ parameter to the list, which you do with a '&' rather than '?' before the parameter name. So, with a JATOS link it will look like this:

ht<span>tps://</span>mywonderfulexperiment.com/publix/1/start?batchId=1&generalMultiple<b>&condition=5</b>

Then, in your experiment script, you want to access the value of the URL parameter called 'condition'. You can either do this using the jsPsych.data.getURLVariable function, as Josh suggested:

`var condition = jsPsych.data.getURLVariable('condition');`

or you can use the [jatos.urlQueryParameters function](http://www.jatos.org/jatos.js-Reference.html#original-url-query-parameters):

`var condition = jatos.urlQueryParameters.condition;`

Then, as Josh said above, you can use the value of this condition variable to set up the timeline accordingly.

Is this all that you're looking for, or did you also want help with the block randomization?

**Reply by LangDevLab** [reply]
Ah, thank you! I managed to get it all to work by wrapping  
`var condition = jatos.urlQueryParameters.condition;`and the rest of the experiment in a jatos wrapper!

Thank you for all your help! :)

**Reply by becky-gilbert** [reply]
Great! Glad it's working. And sorry that I forgot to mention that any `jatos.xxx` functions have to go inside of the `jatos.onLoad` function.

Also, you mentioned that you were wondering if the initial URL redirection that happens at the start of a JATOS-hosted study might cause problems for grabbing the URL parameters. This doesn't cause problems for the `jatos.urlQueryParameters` function, but it might mean that jsPsych's version of this function, `jsPsych.data.getURLVariable`, doesn't work with JATOS studies. Thought I'd mention this since I said in my previous response that you can use either function. I haven't tested this yet and will update my response when I have.
